image/image.go

```diff
@@ -0,0 +1,386 @@
+		if snapName == model.Kernel() || snapName == model.Core() {
+			if err := osutil.CopyFile(fn, filepath.Join(dirs.SnapBlobDir, filepath.Base(fn)), 0); err != nil {
+				return err
+			}
+		}
+		// set seed.yaml
+		seedYaml.Snaps = append(seedYaml.Snaps, &snap.SeedSnap{
+			Name:        info.Name(),
+			SnapID:      info.SnapID,
+			Revision:    info.Revision,
+			Channel:     info.Channel,
+			DeveloperID: info.DeveloperID,
+			Developer:   info.Developer,
+			File:        filepath.Base(fn),
+		})
+	}
+	if err := seedYaml.Write(); err != nil {
+		return fmt.Errorf("cannot write seed.yaml: %s", err)
+	}
+
+	// now do the bootloader stuff
+	if err := partition.InstallBootConfig(opts.GadgetUnpackDir); err != nil {
+		return err
+	}
+
+	if err := setBootvars(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setBootvars() error {
+	// set the bootvars for kernel/os snaps so that the system
+	// actually boots and can do the `firstboot` import of the snaps.
+	//
+	// there is also no mounted os/kernel snap, all we have are the
+	// blobs
+
+	bootloader, err := partition.FindBootloader()
+	if err != nil {
+		return fmt.Errorf("can not set kernel/os bootvars: %s", err)
+	}
+
+	snaps, _ := filepath.Glob(filepath.Join(dirs.SnapBlobDir, "*.snap"))
+	if len(snaps) == 0 {
+		return fmt.Errorf("internal error: cannot find os/kernel snap")
+	}
+	for _, fullname := range snaps {
+		bootvar := ""
+
+		// detect type
+		snapFile, err := snap.Open(fullname)
+		if err != nil {
+			return fmt.Errorf("can not read %v", fullname)
+		}
+		// read .sideinfo
+		var si snap.SideInfo
+		siFn := fullname + ".sideinfo"
+		if osutil.FileExists(siFn) {
+			j, err := ioutil.ReadFile(siFn)
+			if err != nil {
+				return err
+			}
+			if err := json.Unmarshal(j, &si); err != nil {
+				return fmt.Errorf("cannot read metadata: %s %s\n", siFn, err)
+			}
+		}
+		info, err := snap.ReadInfoFromSnapFile(snapFile, &si)
+		if err != nil {
+			return fmt.Errorf("can not get info for %v", fullname)
+		}
+		// local install
+		if info.Revision.Unset() {
+			info.Revision = snap.R(-1)
+		}
+
+		switch info.Type {
+		case snap.TypeOS:
+			bootvar = "snap_core"
+		case snap.TypeKernel:
+			bootvar = "snap_kernel"
+			if err := extractKernelAssets(fullname, info); err != nil {
+				return err
+			}
+		}
+
+		name := filepath.Base(fullname)
+		if bootvar != "" {
+			if err := bootloader.SetBootVar(bootvar, name); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func runCommand(cmdStr ...string) error {
+	cmd := exec.Command(cmdStr[0], cmdStr[1:]...)
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("cannot run %v: %s (%s)", cmdStr, err, output)
+	}
+	return nil
+}
+
+func extractKernelAssets(snapPath string, info *snap.Info) error {
+	// FIXME: hrm, hrm, we need to be root for this - alternatively
+	//        we could make boot.ExtractKernelAssets() work on
+	//        a plain .snap file again (i.e. revert bee59a2)
```

Comment:
Thanks, I will do that in a separate branch to avoid making this review even more work than it already is.